man/RNAfold.1.in
.\" -*-nroff-*-
.\" .ER
.TH "RNAFOLD" "l" "1.6" "Ivo Hofacker" "ViennaRNA"
.SH "NAME"
RNAfold \- calculate secondary structures of RNAs
.SH "SYNOPSIS"
\fBRNAfold\fP [\-p[\fI0|2\fP]] [\-C] [\-T\ \fItemp\fP] [\-4] [\-d[0|1|2|3]]
[\-noLP] [\-noGU] [\-noCloseGU] [\-e\ \fI1|2\fP] [\-P\ \fIparamfile\fP]
[\-nsp\ \fIpairs\fP] [\-S\ \fIscale\fP] [-circ] [-MEA [gamma]]

.SH "DESCRIPTION"
.I RNAfold
reads RNA sequences from stdin, calculates their minimum free energy
(mfe) structure and prints to stdout the mfe structure in bracket
notation and its free energy. If the \-p option was given it also
computes the partition function (pf) and base pairing probability
matrix, and prints the free energy of the thermodynamic ensemble, the
frequency of the mfe structure in the ensemble, and the ensemble
diversity to stdout.

It also produces PostScript files with plots of the resulting
secondary structure graph and a "dot plot" of the base pairing matrix.
The dot plot shows a matrix of squares with area proportional to the
pairing probability in the upper right half, and one square for each pair in
the minimum free energy structure in the lower left half. For each pair
i\-j with probability p>10E\-6 there is a line of the form
.br
i  j  sqrt(p)  ubox
.br
in the PostScript file, so that the pair probabilities can be easily
extracted.

Sequences are read in a simple text format where each sequence occupies
a single line. Each sequence may be preceded by a line of the form
.br
> name
.br
to assign a name to the sequence. If a name is given in the input
PostScript files "name_ss.ps" and "name_dp.ps" are produced for
the structure and dot plot, respectively. Otherwise the file names
default to rna.ps and dot.ps. Existing files of the same name will be
overwritten.
.br
The input format is similar to fasta except that even long sequences
may not be interrupted by line breaks, and the header lines are optional.
The program will continue to read new sequences until a line consisting
of the single character @ or an end of file condition is encountered.
.SH "OPTIONS"
.TP
.B \-p
Calculate the partition function and base pairing probability matrix in
addition to the mfe structure. Default is calculation of mfe structure
only. In addition to the MFE structure we print a coarse representation of
the pair probabilities in form of a pseudo bracket notation, followed by
the ensemble free energy, as well as the centroid structure derived from
the pair probabilities together with its free energy and distance to the
ensemble. Finally it prints the frequency of the mfe structure, and the
structural diversity (mean distance between the structures in the ensemble).
See the description of pf_fold() and mean_bp_dist() and centroid() in the
RNAlib documentation for details.
.br
Note that unless you also specify \-d2 or \-d0, the partition
function and mfe calculations will use a slightly different energy
model. See the discussion of dangling end options below.
.TP
.B \-p0
Calculate the partition function but not the pair probabilities,
saving about 50% in runtime. Prints the ensemble free energy \-kT ln(Z).
.TP
.B \-p2
In addition to pair probabilities compute stack probabilities, i.e. the
probability that a pair (i,j) and the immediately interior pair (i+1,j-1) are
formed simultaneously. A second postscript dot plot called "name_dp2.ps", or
"dot2.ps" (if the sequence does not have a name), is produced that contains
pair probabilities in the upper right half and stack probabilities in the
lower left.
.TP
.BI \-MEA\  [gamma]
Calculate an MEA (maximum expected accuracy) structure, where the expected
accuracy is computed from the pair probabilities: each base pair (i,j) gets
a score 2*gamma*p_ij and the score of an unpaired base is given by the
probability of not forming a pair. The parameter gamma tunes the importance
of correctly predicted pairs versus unpaired bases. Thus, for small values of
gamma the MEA structure will contain only pairs with very high probability. 
The default value is gamma=1.
Using -MEA implies -p for computing the pair probabilities.
.TP
.B \-C
Calculate structures subject to constraints.
The program reads first the sequence, then a string containing constraints
on the structure encoded with the symbols:
. (no constraint for this base)
| (the corresponding base has to be paired
x (the base is unpaired)
< (base i is paired with a base j>i)
> (base i is paired with a base j<i)
and matching brackets ( ) (base i pairs base j)
With the exception of "|", constraints will disallow all pairs conflicting
with the constraint. This is usually sufficient to enforce the constraint,
but occasionally a base may stay unpaired in spite of constraints. PF
folding ignores constraints of type "|".
.TP
.B \-T\ \fItemp\fP
Rescale energy parameters to a temperature of \fItemp\fP C. Default is 37C.
.TP
.B \-4
Do not include special stabilizing energies for certain tetra\-loops. Mostly
for testing.
.TP
.B \-d[0|1|2|3]
How to treat "dangling end" energies for bases adjacent to helices in
free ends and multi\-loops: With (\-d1) only unpaired bases can
participate in at most one dangling end, this is the default for mfe
folding but unsupported for the partition function folding. With
\fB\-d2\fP this check is ignored, dangling energies will be added for
the bases adjacent to a helix on both sides in any case; this is the
default for partition function folding (\-p). \fB\-d\fP or \fB\-d0\fP
ignores dangling ends altogether (mostly for debugging).
.br
With \fB\-d3\fP mfe folding will allow coaxial stacking of adjacent helices
in multi\-loops. At the moment the implementation will not allow coaxial
stacking of the two interior pairs in a loop of degree 3 and works
only for mfe folding.
.br
Note that by default (as well as with \-d1 and \-d3) pf and mfe
folding treat dangling ends differently. Use \fB\-d2\fR in addition to
\fB\-p\fR to ensure that both algorithms use the same energy model.
.TP
.B \-noLP
Produce structures without lonely pairs (helices of length 1).
For partition function folding this only disallows pairs that can
\fBonly\fP occur isolated. Other pairs may still occasionally
occur as helices of length 1.
.TP
.B \-noGU
Do not allow GU pairs.
.TP
.B \-noCloseGU
Do not allow GU pairs at the end of helices.
.TP
.B \-e\ \fI1|2\fP
Rarely used option to fold sequences from the artificial ABCD... alphabet,
where A pairs B, C\-D etc.  Use the energy parameters for GC (\-e 1) or AU
(\-e 2) pairs.
.TP
.B \-P\ \fI<paramfile>\fP
Read energy parameters from \fIparamfile\fP, instead of using the default
parameter set. A sample parameter file should accompany your distribution.
See the RNAlib documentation for details on the file format.
.TP
.B \-nsp  \fIpairs\fP
Allow other pairs in addition to the usual AU,GC,and GU pairs. \fIpairs\fP
is a comma separated list of additionally allowed pairs. If a the first
character is a "\-" then AB will imply that AB and BA are allowed pairs.
e.g. RNAfold \-nsp \-GA  will allow GA and AG pairs. Nonstandard pairs are
given 0 stacking energy.
.TP
.B \-S\ \fIscale\fP
In the calculation of the pf use \fIscale\fP*mfe as an estimate for the
ensemble free energy (used to avoid overflows). The default is 1.07,
useful values are 1.0 to 1.2. Occasionally needed for long sequences.
.TP
.B \-circ
Assume a circular (instead of linear) RNA molecule.
.TP
.B \-noPS
Do not produce postscript drawing of the mfe structure.

.SH "REFERENCES"
The calculation of mfe structures is based on dynamic
programming algorithm originally developed by M. Zuker and P. Stiegler.
The partition function algorithm is based on work by J.S. McCaskill.
The energy parameters are taken from:
.br
D.H. Mathews, J. Sabina, M. Zuker and H. Turner
"Expanded Sequence Dependence of Thermodynamic Parameters Provides
Robust Prediction of RNA Secondary Structure"
JMB, 288, pp 911\-940, 1999
.br
A. Walter, D Turner, J Kim, M Lyttle, P M\[:u]ller, D Mathews, M Zuker
"Coaxial stacking of helices enhances binding of oligoribonucleotides.."
PNAS, 91, pp 9218\-9222, 1994
.PP
If you use this program in your work you might want to cite:
.PP
I.L. Hofacker, W. Fontana, P.F. Stadler, S. Bonhoeffer, M. Tacker, P. Schuster
(1994)
Fast Folding and Comparison of RNA Secondary Structures.
Monatshefte f. Chemie 125: 167\-188
.br
M. Zuker, P. Stiegler (1981) Optimal computer folding of large RNA
sequences using thermodynamic and auxiliary information, Nucl Acid Res
9: 133\-148
.br
J.S. McCaskill (1990) The equilibrium partition function and base pair
binding probabilities for RNA secondary structures, Biopolymers 29: 1105\-1119
.br
I.L. Hofacker & P.F. Stadler (2006)
Memory Efficient Folding Algorithms for Circular RNA Secondary
Structures, Bioinformatics
.br
A.F. Bompfünewerer, R. Backofen, S.H. Bernhart, J. Hertel, I.L. Hofacker,
P.F. Stadler, S. Will (2007)
"Variations on RNA Folding and Alignment: Lessons from Benasque"
J. Math. Biol.
.br
D. Adams (1979) The hitchhiker's guide to the galaxy, Pan Books, London
.SH "VERSION"
This man page documents version @VERSION@ Vienna RNA Package.
.SH "AUTHORS"
Ivo L Hofacker, Walter Fontana, Sebastian Bonhoeffer, Peter F Stadler.
.SH "BUGS"
If in doubt our program is right, nature is at fault.
Comments should be sent to rna@tbi.univie.ac.at.


NEWS
See Changelog for details.

Version 2.0
------------
- yeeeehaa, its almost done...

- All algorithms use the Turner'04 nearest neighbor model
- The RNAlib provides (OpenMP) threadsafe folding routines per default. This is
  enables concurrent calls to the folding routines in parallel. The feature
  can be disabled by passing '--disable-openmp' to the configure script
- serious changes in command line parameters. Everything complies with
  GNU standard from now on (short options with preceding '-', long options
  with preceding '--'.
- FASTA file support for RNAfold. RNA sequences do not need to be passed
  on a single line anymore when a FASTA header is provided.
- The new program RNA2Dfold computes MFE, partition function and
  stochastically sampled secondary structures in a partitioning of the secondary
  structure space according to the base pair distance to two reference structures
- The new program PKplex computes...
- The new program RNALfoldz computes locally stable secondary structures
  together with a z-score
- The new program RNALalifold computes locally stable consensus structures for
  alignments
- The new program RNAparconv enables the conversion of 'old' energy parameter files
  (v1.4-v1.8) to the new format used in version 2.x


Version 1.8
------------
- new RNAalifold has better treatment of gaps and ribosum based covariance
  scores. Use the -old switch for compatibility with older RNAalifold
  versions.
- RNAplfold -u  now computes all accessibilities up to a maximum length
  (much faster than computing each individually)
- ATTENTION: output formats of RNAplfold -u and or RNAup have been changed  
  Programs parsing RNAplfold and RNAup output will have to be modified.
- RNAfold and RNAalifold compute centroid structures when run with -p
  use the  -MEA  option to compute Maximum Expected Accuracy structures.

Version 1.7
------------
- RNAplfold can now be used to compute accessibilities, i.e. the
  probability that a stretch of the RNA remains unpaired (and thus
  available for intermolecular interactions).
- A new version of RNAup predicts RNA-RNA interactions takeing into account
  the competition between inter- and intramolecular structure in both
  molcules
- Circular RNAs can be treated by RNAfold, RNAalifold, RNAsubopt, and RNAcofold
- RNAaliduplex predicts RNA-RNA interactions between two sets of aligned
  sequences (inter-molecular structure only)

Version 1.6
------------
- The RNAforester program for tree-alignments of RNA structures is now
  distributed with the Vienna RNA package, see the RNAforester
  subdirectory for more information. RNAforester was written by Matthias
  Hoechsmann <mhoechsm@techfak.uni-bielefeld.de> 
- The Kinfold program for stochastic simulation of folding trajectories is
  now included in the package, see the Kinfold subdirectory.
- cofolding of two structures now supports suboptimal folding and 
  partition function folding. 
  ATTENTION: Energies of hybrid structures now include the Duplex-initiation
  energy, which was neglected in previous version.
- RNAplfold is a partition function variant of RNALfold. It computes the
  mean probability of a (local) base pair averaged over all sequence
  windows that contain the pair.
- new utilities to color alignments and consensus structures  
- RNAfold -p now computes the centroid structure
- ATTENTION: ensemble diversities in version <1.6.5 are off by a factor 2

Version 1.5pre 
-----------
- ViennaRNA now uses autoconfig generated configure scripts for even better
  portability (should compile on any UNIX, Linux, MacOS X, Windows with
  Cygwin).
- The new RNAalifold program predicts consensus structures for a set of
  aligned sequences.
- Complete suboptimal folding is now integrated in the library.
- Beginning support for co-folding of two strands: energy_of_struct() and
  RNAeval can now compute energies of duplex structures.
- RNAcofold predicts hybrid structures of two RNA strands
- RNAduplex predicts hybrid structures, while allowing only inter-molecular
  base pairs (useful for finding potential binding sites)
- RNALfold predicts locally stable structures in long sequences.
- Major changes to Perl module. See the pod documentation (perldoc RNA).
- RNAsubopt can do stochastic backtracking to produce samples of suboptimal
  structures with Boltzmann statistics.
- New utilities to rotate secondary structure plots and annotate them with
  reliability data.
- Various small bug fixes

Version 1.4
-----------
- New Turner parameters as described in Mathews et.al. JMB v288, 1999.
  Small changes to format of parameter files (old param files won't work!)
- mfe and suboptimal folding will produce only structures without isolated
  pairs if noLonelyPairs=1 (-noLP option), for partition function folding
  pairs that can only occur as isolated pairs are not formed.
- setting dangles=3 (-d3 option) will allow co-axial stacking of adjacent
  helices in mfe folding and energy_of_struct().

Version 1.3.1
-------------
- RNAheat would produce spikes in the specific heat because dangling end
  energies did not go smoothly to 0. 
- PS dot plots now have an option to use a log scale (edit _dp.ps file
  and set logscale to true).

Version 1.3
-----------
- Secondary structure plots now use E. Bruccoleri's naview routines for
  layout by default. New utility RNAplot produces secondary structure plots 
  from structures in bracket notation with several options.
- New -d2 option in RNAfold and RNAeval sets dangles=2, which makes 
  energy_of_struct() and fold() treat dangling ends as in pf_fold().
  -noLP option in RNAfold etc sets noLonelyPairs=1, which avoids most
  structures containing lonely base pairs (helices of length 1).
- new utility functions pack_structure() unpack_structure() make_pair_table()
  and bp_distance().  RNAdistance adds bp_distance() via -DP switch.
- First release of RNAsubopt for complete suboptimal folding.
- fixed bug in asymmetry penalty for interior loops. 

Progs/RNAfold.c
/* Last changed Time-stamp: <2007-12-05 13:55:42 ronny> */
/*
                  Ineractive Access to folding Routines

                  c Ivo L Hofacker
                  Vienna RNA package
*/

/** \file
*** \brief RNAfold program source code
***
*** This code provides an interface for MFE and Partition function folding
*** of single linear or circular RNA molecules.
**/

#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <ctype.h>
#include <unistd.h>
#include <string.h>
#include "fold.h"
#include "part_func.h"
#include "fold_vars.h"
#include "PS_dot.h"
#include "utils.h"
#include "read_epars.h"
#include "MEA.h"
#include "RNAfold_cmdl.h"

/*@unused@*/
static char UNUSED rcsid[] = "$Id: RNAfold.c,v 1.25 2009/02/24 14:22:21 ivo Exp $";

/*--------------------------------------------------------------------------*/

int main(int argc, char *argv[]){
  struct        RNAfold_args_info args_info;
  char          *buf, *rec_sequence, *rec_id, **rec_rest, *structure=NULL, *cstruc=NULL;
  char          fname[80], ffname[80], *ParamFile=NULL;
  char          *ns_bases=NULL, *c;
  int           i, length, l, cl, sym, r, istty, pf, noPS, noconv, fasta;
  unsigned int  rec_type, read_opt;
  double        energy, min_en, kT, sfact;
  int           doMEA=0, circular;
  double        MEAgamma = 1.;

  rec_type      = read_opt = 0;
  rec_id        = rec_sequence = NULL;
  rec_rest      = NULL;
  do_backtrack  = 1;
  pf            = 0;
  sfact         = 1.07;
  noPS          = 0;
  noconv        = 0;
  circular      = 0;
  fasta         = 0;
  cl            = l = length = 0;
  dangles	= 2;

  /*
  #############################################
  # check the command line parameters
  #############################################
  */
  if(RNAfold_cmdline_parser (argc, argv, &args_info) != 0) exit(1);
  /* temperature */
  if(args_info.temp_given)        temperature = args_info.temp_arg;
  /* structure constraint */
  if(args_info.constraint_given)  fold_constrained=1;
  /* do not take special tetra loop energies into account */
  if(args_info.noTetra_given)     tetra_loop=0;
  /* set dangle model */
  if(args_info.dangles_given)     dangles = args_info.dangles_arg;
  /* do not allow weak pairs */
  if(args_info.noLP_given)        noLonelyPairs = 1;
  /* do not allow wobble pairs (GU) */
  if(args_info.noGU_given)        noGU = 1;
  /* do not allow weak closing pairs (AU,GU) */
  if(args_info.noClosingGU_given) no_closingGU = 1;
  /* do not convert DNA nucleotide "T" to appropriate RNA "U" */
  if(args_info.noconv_given)      noconv = 1;
  /* set energy model */
  if(args_info.energyModel_given) energy_set = args_info.energyModel_arg;
  /* take another energy parameter set */
  if(args_info.paramFile_given)   ParamFile = strdup(args_info.paramFile_arg);
  /* Allow other pairs in addition to the usual AU,GC,and GU pairs */
  if(args_info.nsp_given)         ns_bases = strdup(args_info.nsp_arg);
  /* set pf scaling factor */
  if(args_info.pfScale_given)     sfact = args_info.pfScale_arg;
  /* assume RNA sequence to be circular */
  if(args_info.circ_given)        circular=1;
  /* do not produce postscript output */
  if(args_info.noPS_given)        noPS=1;
  /* partition function settings */
  if(args_info.partfunc_given){
    pf = 1;
    if(args_info.partfunc_arg != -1)
      do_backtrack = args_info.partfunc_arg;
  }
  /* MEA (maximum expected accuracy) settings */
  if(args_info.MEA_given){
    pf = doMEA = 1;
    if(args_info.MEA_arg != -1)
      MEAgamma = args_info.MEA_arg;
  }

  /* free allocated memory of command line data structure */
  RNAfold_cmdline_parser_free (&args_info);

  /*
  #############################################
  # begin initializing
  #############################################
  */
  if (ParamFile != NULL)
    read_parameter_file(ParamFile);

  if (circular && noLonelyPairs)
    warn_user("depending on the origin of the circular sequence, some structures may be missed when using -noLP\nTry rotating your sequence a few times");

  if (ns_bases != NULL) {
    nonstandards = space(33);
    c=ns_bases;
    i=sym=0;
    if (*c=='-') {
      sym=1; c++;
    }
    while (*c!='\0') {
      if (*c!=',') {
        nonstandards[i++]=*c++;
        nonstandards[i++]=*c;
        if ((sym)&&(*c!=*(c-1))) {
          nonstandards[i++]=*c;
          nonstandards[i++]=*(c-1);
        }
      }
      c++;
    }
  }

  istty = isatty(fileno(stdout))&&isatty(fileno(stdin));

  /* print user help if we get input from tty */
  if(istty){
    if(fold_constrained){
      print_tty_constraint_full();
      print_tty_input_seq_str("Input sequence (upper or lower case) followed by structure constraint");
    }
    else print_tty_input_seq();
  }

  /* set options we wanna pass to read_record */
  if(istty)             read_opt |= VRNA_INPUT_NOSKIP_BLANK_LINES;
  if(!fold_constrained) read_opt |= VRNA_INPUT_NO_REST;

  /*
  #############################################
  # main loop: continue until end of file
  #############################################
  */
  while(
    !((rec_type = read_record(&rec_id, &rec_sequence, &rec_rest, read_opt))
        & (VRNA_INPUT_ERROR | VRNA_INPUT_QUIT))){

    /*
    ########################################################
    # init everything according to the data we've read
    ########################################################
    */
    if(rec_id){
      if(!istty) printf("%s\n", rec_id);
      (void) sscanf(rec_id, ">%42s", fname);
    }
    else fname[0] = '\0';

    length  = (int)strlen(rec_sequence);

    /* parse the rest of the current dataset to obtain a structure constraint */
    if(fold_constrained){
      if(rec_rest)
        for(r=0,i=0;rec_rest[i];i++){
          l = (int)strlen(rec_rest[i]);
          c = (char *) space(sizeof(char) * (l+1));
          (void) sscanf(rec_rest[i], "%s", c);
          /* actual constraining structure ? */
          if((*c=='<') || (*c=='>') || (*c=='.') || (*c=='|') || (*c=='(') || (*c==')') || (*c=='x')){
            r += l+1;
            cstruc = (char *)xrealloc(cstruc, r*sizeof(char));
            strcat(cstruc, c);
            if(!rec_id || ((int)strlen(c) != l)) break; /* stop if not in fasta mode or multiple words on line */
          }
          else if(r) break; /* we skip only leading comments and other stuff */
        }
      cl = (cstruc) ? (int)strlen(cstruc) : 0;
      if(cl == 0)           warn_user("structure constraint is missing");
      else if(cl < length)  warn_user("structure constraint is shorter than sequence");
      else if(cl > length)  nrerror("structure constraint is too long");
    }
    structure = (char *)space(sizeof(char) *(length+1));
    if(fold_constrained && (cstruc)) strncpy(structure, cstruc, sizeof(char)*(cl+1));

    if(noconv)  str_RNA2RNA(rec_sequence);
    else        str_DNA2RNA(rec_sequence);

    if(istty) printf("length = %d\n", length);

    /*
    ########################################################
    # begin actual computations
    ########################################################
    */
    min_en = (circular) ? circfold(rec_sequence, structure) : fold(rec_sequence, structure);

    printf("%s\n%s", rec_sequence, structure);
    if (istty)
      printf("\n minimum free energy = %6.2f kcal/mol\n", min_en);
    else
      printf(" (%6.2f)\n", min_en);

    (void) fflush(stdout);

    if(fname[0] != '\0'){
      strcpy(ffname, fname);
      strcat(ffname, "_ss.ps");
    } else strcpy(ffname, "rna.ps");

    if (!noPS) (void) PS_rna_plot(rec_sequence, structure, ffname);
    if (length>2000) free_arrays();
    if (pf) {
      char *pf_struc;
      pf_struc = (char *) space((unsigned) length+1);
      if (dangles==1) {
          dangles=2;   /* recompute with dangles as in pf_fold() */
          min_en = (circular) ? energy_of_circ_structure(rec_sequence, structure, 0) : energy_of_structure(rec_sequence, structure, 0);
          dangles=1;
      }

      kT = (temperature+273.15)*1.98717/1000.; /* in Kcal */
      pf_scale = exp(-(sfact*min_en)/kT/length);
      if (length>2000) fprintf(stderr, "scaling factor %f\n", pf_scale);

      if (cstruc!=NULL) strncpy(pf_struc, cstruc, length+1);

      energy = (circular) ? pf_circ_fold(rec_sequence, pf_struc) : pf_fold(rec_sequence, pf_struc);

      if (do_backtrack) {
        printf("%s", pf_struc);
        if (!istty) printf(" [%6.2f]\n", energy);
        else printf("\n");
      }
      if ((istty)||(!do_backtrack))
        printf(" free energy of ensemble = %6.2f kcal/mol\n", energy);
      if (do_backtrack) {
        plist *pl1,*pl2;
        char *cent;
        double dist, cent_en;
        FLT_OR_DBL *probs = export_bppm();
        assign_plist_from_pr(&pl1, probs, length, 1e-5);
        assign_plist_from_db(&pl2, structure, 0.95*0.95);
        /* cent = centroid(length, &dist); <- NOT THREADSAFE */
        cent = get_centroid_struct_pr(length, &dist, pr);
        cent_en = (circular) ? energy_of_circ_structure(rec_sequence, cent, 0) :energy_of_structure(rec_sequence, cent, 0);
        printf("%s {%6.2f d=%.2f}\n", cent, cent_en, dist);
        free(cent);
        if (fname[0]!='\0') {
          strcpy(ffname, fname);
          strcat(ffname, "_dp.ps");
        } else strcpy(ffname, "dot.ps");
        (void) PS_dot_plot_list(rec_sequence, ffname, pl1, pl2, "");
        free(pl2);
        if (do_backtrack==2) {
          pl2 = stackProb(1e-5);
          if (fname[0]!='\0') {
            strcpy(ffname, fname);
            strcat(ffname, "_dp2.ps");
          } else strcpy(ffname, "dot2.ps");
          PS_dot_plot_list(rec_sequence, ffname, pl1, pl2,
                           "Probabilities for stacked pairs (i,j)(i+1,j-1)");
          free(pl2);
        }
        free(pl1);
        free(pf_struc);
        if(doMEA){
          float mea, mea_en;
          plist *pl;
          assign_plist_from_pr(&pl, pr, length, 1e-4/(1+MEAgamma));
          mea = MEA(pl, structure, MEAgamma);
          mea_en = (circular) ? energy_of_circ_structure(rec_sequence, structure, 0) : energy_of_structure(rec_sequence, structure, 0);
          printf("%s {%6.2f MEA=%.2f}\n", structure, mea_en, mea);
          free(pl);
        }
      }
      printf(" frequency of mfe structure in ensemble %g; ", exp((energy-min_en)/kT));
      if (do_backtrack)
        printf("ensemble diversity %-6.2f", mean_bp_distance(length));
      printf("\n");

      free_pf_arrays();
    }
    (void) fflush(stdout);

    /* clean up */
    if(cstruc) free(cstruc);
    if(rec_id) free(rec_id);
    free(rec_sequence);
    free(structure);
    /* free the rest of current dataset */
    if(rec_rest){
      for(i=0;rec_rest[i];i++) free(rec_rest[i]);
      free(rec_rest);
    }
    rec_id = rec_sequence = structure = cstruc = NULL;
    rec_rest = NULL;

    /* print user help for the next round if we get input from tty */
    if(istty){
      if(fold_constrained){
        print_tty_constraint_full();
        print_tty_input_seq_str("Input sequence (upper or lower case) followed by structure constraint");
      }
      else print_tty_input_seq();
    }
  }
  return EXIT_SUCCESS;
}
